<a href="https://colab.research.google.com/github/jeselsohn470/jeselsohn470.github.io/blob/master/hw_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.

In [ ]:
import math
import pandas as pd
import numpy as np

Make an untagged vector, point, and two tagged vectors, data1 and data2:

In [ ]:
point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']


Write code to separate the data (X) from the tag (Y).  Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F


In [ ]:
print("The vector " , data1[0:-1], " has tag ", data1[-1])

The vector  [1, 1, 1]  has tag  M


Now let's classify the point as either M or F.  We'll do this by setting k = 1 and using the Euclidean distance.  We'll define that as: 

In [ ]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)


Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance.

In [ ]:
print(euclideanDistance(data1,point,3))

1.4142135623730951


Now, let's get more data from a file, myFile.csv. This code will read it for us:

In [ ]:
url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0, error_bad_lines=False) 
#put data in dataset without header line
dataset = np.array(df)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


Show that you understand what you read by:

1. Printing the first two vectors in the file
2. Printing the Euclidean distance between those two vectors

Here is some code which I think might help get you on your way!

In [ ]:
print(len(dataset))
print(euclideanDistance(dataset[0],dataset[1],3))

10
5.744562646538029


In [ ]:
print(dataset[0])
print(dataset[1])
print(euclideanDistance(dataset[0],dataset[1],3))

[0 1 2 'F']
[1 5 6 'F']
5.744562646538029


Now assume you have a new value for point:

[0,0,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:  

We suggest defining some type of data struction to store different vectors' distances and their tags like this:

In [ ]:
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point


You can then add vector distances like this:

In [ ]:
eucDistances = [] # list of distances, will hold objects of type distClass

temp=dataset[1]
label=temp[-1]
d=euclideanDistance(point,temp,3)
print("The distances between " , point , " and " , temp,  " is " , str(d))
print(" and the label is " + label)
obj = distClass() #one record's distance and tag
obj.dist=d
obj.tag=label
eucDistances.append(obj)


The distances between  [1, 0, 0, '?']  and  [1 5 6 'F']  is  7.810249675906654
 and the label is F


In [ ]:
eucDistances[0].dist

7.810249675906654

and sort them like this:

In [ ]:
eucDistances.sort(key=lambda x: x.dist) 

# **Euclidean Distance**

In [ ]:
eucDistances = []
point = [0,0,100]
for i in range(len(dataset)):
  d=euclideanDistance(point, dataset[i],3) 
  label= dataset[i][-1]
  obj = distClass() #one record's distance and tag
  obj.dist=d
  obj.tag=label
  eucDistances.append(obj)

In [ ]:
eucDistances.sort(key=lambda x: x.dist) 
for i in range(len(eucDistances)):
  print('The distance is {} and the gender is {}.'.format(eucDistances[i].dist, eucDistances[i].tag))


The distance is 80.00624975587844 and the gender is M.
The distance is 94.13819628609845 and the gender is F.
The distance is 94.13819628609845 and the gender is M.
The distance is 98.00510190801293 and the gender is F.
The distance is 98.04080783020915 and the gender is F.
The distance is 99.08582138732059 and the gender is M.
The distance is 100.45894683899488 and the gender is F.
The distance is 100.68266980965493 and the gender is M.
The distance is 100.91580649234291 and the gender is M.
The distance is 106.78014796768171 and the gender is M.


In [ ]:
male = 0
female = 0
k = [1, 3]
for k_num in range(len(k)):
  for i in range(0, k[k_num]):
    gender = eucDistances[i].tag
    if gender is 'M':
      male = male + 1
    else:
      female = female + 1
  print("Using {} for K, the result is the following:".format(k[k_num]))
  if male > female:
    print("The Gender of this text is MALE.")
  else:
    print("The Gender of this text is FEMALE.")
  print()
  male = 0
  female = 0

Using 1 for K, the result is the following:
The Gender of this text is MALE.

Using 3 for K, the result is the following:
The Gender of this text is MALE.



Questions:

1. What is the label for point if k=1?
2. What is the label for point if k=3?
3. Would the result be different if we used a different distance function like Hamming or Manhattan?

In [ ]:
#Add code with functions that implement Hamming and Manhattan distances and test 
#what the label will be for k=1 and k=3 for all possibilities 
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)


# **Manhattan Distance**

In [ ]:
def manhattanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + abs(num1-num2)
   return distance

In [ ]:
manDistances = []
point = [0,0,100]
for i in range(len(dataset)):
  d = manhattanDistance(point, dataset[i],3) 
  label= dataset[i][-1]
  obj = distClass() #one record's distance and tag
  obj.dist=d
  obj.tag=label
  manDistances.append(obj)

In [ ]:
manDistances.sort(key=lambda x: x.dist) 
for i in range(len(eucDistances)):
  print('The distance is {} and the gender is {}.'.format(manDistances[i].dist, manDistances[i].tag))

The distance is 81.0 and the gender is M.
The distance is 99.0 and the gender is F.
The distance is 100.0 and the gender is F.
The distance is 100.0 and the gender is M.
The distance is 102.0 and the gender is F.
The distance is 104.0 and the gender is M.
The distance is 122.0 and the gender is F.
The distance is 123.0 and the gender is M.
The distance is 124.0 and the gender is M.
The distance is 140.0 and the gender is M.


In [ ]:
male = 0
female = 0
k = [1, 3]
for k_num in range(len(k)):
  for i in range(0, k[k_num]):
    gender = manDistances[i].tag
    if gender is 'M':
      male = male + 1
    else:
      female = female + 1
  print("Using {} for K, the result is the following:".format(k[k_num]))
  if male > female:
    print("The Gender of this text is MALE.")
  else:
    print("The Gender of this text is FEMALE.")
  print()
  male = 0
  female = 0

Using 1 for K, the result is the following:
The Gender of this text is MALE.

Using 3 for K, the result is the following:
The Gender of this text is FEMALE.



# **Hamming Distance**

In [ ]:
def hammingDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         if num1 != num2:
          distance = distance + 1
   return distance

In [ ]:
hamDistances = []
point = [0,0,100]
for i in range(len(dataset)):
  d = hammingDistance(point, dataset[i],3) 
  label= dataset[i][-1]
  obj = distClass() #one record's distance and tag
  obj.dist=d
  obj.tag=label
  hamDistances.append(obj)

In [ ]:
hamDistances.sort(key=lambda x: x.dist) 
for i in range(len(eucDistances)):
  print('The distance is {} and the gender is {}.'.format(hamDistances[i].dist, hamDistances[i].tag))

The distance is 2 and the gender is F.
The distance is 2 and the gender is M.
The distance is 3 and the gender is F.
The distance is 3 and the gender is M.
The distance is 3 and the gender is F.
The distance is 3 and the gender is M.
The distance is 3 and the gender is M.
The distance is 3 and the gender is F.
The distance is 3 and the gender is M.
The distance is 3 and the gender is M.


In [ ]:
male = 0
female = 0
k = [1, 3]
for k_num in range(len(k)):
  for i in range(0, k[k_num]):
    gender = hamDistances[i].tag
    if gender is 'M':
      male = male + 1
    else:
      female = female + 1
  print("Using {} for K, the result is the following:".format(k[k_num]))
  if male > female:
    print("The Gender of this text is MALE.")
  else:
    print("The Gender of this text is FEMALE.")
  print()
  male = 0
  female = 0

Using 1 for K, the result is the following:
The Gender of this text is FEMALE.

Using 3 for K, the result is the following:
The Gender of this text is FEMALE.



Up until here is a simplified version of the homework.
Below here is the work for the part we will be checking as the basis of your grade: 

Now let's look at some bigger files: 

1. mytrain.csv (for training the model)
2. mytest.csv (for testing)


In [ ]:
url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))

In [ ]:
print(train_data.shape)# number of records and features
print(train_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]


In [ ]:
print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:
1. k = 1
2. k = 7
3. k = 15

and at first use the Euclidean distance.
Classify each of the vectors in the test_data dataset using the training data from train_data.  Which value for k did the best?  What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.  

Once you are done, you should have a total of 9 different results:

1. Three results for the different value of k using the Euclidean Distance
2. Three results for the different value of k using the Hamming Distance
3. Three results for the different value of k using the Manhattan Distance

Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!

# **Euclidean Distance**

In [ ]:
#Add code here
# use this array to find the accuracy using all three k values 
k = [1, 7, 15]
for k_num in range(len(k)):
  correctly_identified = 0
  incorrectly_identified = 0
  # go through each test point
  for test_point in test_data:
    eucDistances = []
    # for each test point, get all the distances betweent that point and all the training points
    for i in range(len(train_data)):
      d=euclideanDistance(test_point, train_data[i], 29) 
      label= train_data[i][-1]
      obj = distClass() #one record's distance and tag
      obj.dist=d
      obj.tag=label
      eucDistances.append(obj)
    # sort all the distances
    eucDistances.sort(key=lambda x: x.dist) 
    
    # have these two variables to see which one has the most closest poionts to the test point
    male = 0
    female = 0
    for j in range(0, k[k_num]):
        # find the gender which the algorithm predicts using the k closest points
        gender = eucDistances[j].tag
        if gender is 'M':
          male = male + 1
        else:
          female = female + 1
    # check if the algorithm predicted correctly.

    # if the k closests points were male
    if male >= female:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'M':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
    # if the k closests points were female
    else:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'F':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
  print("Using {} for K, the accuracy using the Euclidean Distance  is the following: {}".format(k[k_num], (correctly_identified / (correctly_identified + incorrectly_identified))))
  print()
  # reset the  variable for the next iteration of k = 
  correctly_identified = 0
  incorrectly_identified = 0
  male = 0
  female = 0

Using 1 for K, the accuracy using the Euclidean Distance  is the following: 0.5

Using 7 for K, the accuracy using the Euclidean Distance  is the following: 0.74

Using 15 for K, the accuracy using the Euclidean Distance  is the following: 0.7



**Analysis**

We see here that for Euclidean Distance, k =7 did the best with an accuracy of .74 percent.

# **Hamming Distance**

In [ ]:
#Add code here
# use this array to find the accuracy using all three k values 
k = [1, 7, 15]
for k_num in range(len(k)):
  correctly_identified = 0
  incorrectly_identified = 0
  # go through each test point
  for test_point in test_data:
    hamDistances = []
    # for each test point, get all the distances betweent that point and all the training points
    for i in range(len(train_data)):
      d=hammingDistance(test_point, train_data[i], 29) 
      label= train_data[i][-1]
      obj = distClass() #one record's distance and tag
      obj.dist=d
      obj.tag=label
      hamDistances.append(obj)
    # sort all the distances
    hamDistances.sort(key=lambda x: x.dist) 
    
    # have these two variables to see which one has the most closest poionts to the test point
    male = 0
    female = 0
    for j in range(0, k[k_num]):
        # find the gender which the algorithm predicts using the k closest points
        gender = hamDistances[j].tag
        if gender is 'M':
          male = male + 1
        else:
          female = female + 1
    # check if the algorithm predicted correctly.

    # if the k closests points were male
    if male >= female:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'M':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
    # if the k closests points were female
    else:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'F':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
  print("Using {} for K, the accuracy using the Hamming Distance  is the following: {}".format(k[k_num], (correctly_identified / (correctly_identified + incorrectly_identified))))
  print()
  # reset the  variable for the next iteration of k = 
  correctly_identified = 0
  incorrectly_identified = 0
  male = 0
  female = 0

Using 1 for K, the accuracy using the Hamming Distance  is the following: 0.61

Using 7 for K, the accuracy using the Hamming Distance  is the following: 0.55

Using 15 for K, the accuracy using the Hamming Distance  is the following: 0.57



**Analysis**

We see here that for Hamming Distance, k =1 did the best with an accuracy of .61 percent.

# **Manhattan Distance**

In [ ]:
#Add code here
# use this array to find the accuracy using all three k values 
k = [1, 7, 15]
for k_num in range(len(k)):
  correctly_identified = 0
  incorrectly_identified = 0
  # go through each test point
  for test_point in test_data:
    manhattanDistances = []
    # for each test point, get all the distances betweent that point and all the training points
    for i in range(len(train_data)):
      d=manhattanDistance(test_point, train_data[i], 29) 
      label= train_data[i][-1]
      obj = distClass() #one record's distance and tag
      obj.dist=d
      obj.tag=label
      manhattanDistances.append(obj)
    # sort all the distances
    manhattanDistances.sort(key=lambda x: x.dist) 
    
    # have these two variables to see which one has the most closest poionts to the test point
    male = 0
    female = 0
    for j in range(0, k[k_num]):
        # find the gender which the algorithm predicts using the k closest points
        gender = manhattanDistances[j].tag
        if gender is 'M':
          male = male + 1
        else:
          female = female + 1
    # check if the algorithm predicted correctly.

    # if the k closests points were male
    if male >= female:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'M':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
    # if the k closests points were female
    else:
        # if the prediced value is  the same as the actual value
        if test_point[len(test_point) - 1] is 'F':
          correctly_identified += 1
        else:
          incorrectly_identified += 1
  print("Using {} for K, the accuracy using the Manhattan Distance  is the following: {}".format(k[k_num], (correctly_identified / (correctly_identified + incorrectly_identified))))
  print()
  # reset the  variable for the next iteration of k = 
  correctly_identified = 0
  incorrectly_identified = 0
  male = 0
  female = 0

Using 1 for K, the accuracy using the Manhattan Distance  is the following: 0.61

Using 7 for K, the accuracy using the Manhattan Distance  is the following: 0.63

Using 15 for K, the accuracy using the Manhattan Distance  is the following: 0.69



**Analysis**

We see here that for Manhattans Distance, k =15 did the best with an accuracy of .69 percent.

# **OVERALL ANALYSIS**

Overall, it seems that that the KNN  algorithm which worked the best was with using the euclidean distance and setting k to 7. This led  to an accuracy of .74.

Grade Basis:

80% for correct answers (and yes, there are possibilities that multiple answers are correct-- especially for cases of ties).

20% : Documentation and easily readable code

Please publish your final Notebook in your Github directory.
